src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -900,6 +900,16 @@ lemma cons_is_cycle {u v : V} (p : G.path v u) (h : G.adj u v)
   (he : ¬ ⟦(u, v)⟧ ∈ (p : G.walk v u).edges) : (walk.cons h ↑p).is_cycle :=
 by simp [walk.is_cycle_def, walk.cons_is_trail_iff, he]
 
+lemma cons_is_cycle_iff {u v : V} (p : G.walk v u) (h : G.adj u v) :
+  (walk.cons h p).is_cycle ↔ p.is_path ∧ ¬ ⟦(u, v)⟧ ∈ p.edges :=
+begin
+  split,
+  { simp only [walk.is_cycle_def, walk.cons_is_trail_iff, ne.def, not_false_iff, walk.support_cons,
+               list.tail_cons, true_and, simple_graph.walk.is_path_def],
+    tauto, },
+  { exact λ ⟨hp,he⟩, cons_is_cycle (⟨p,hp⟩ : G.path v u) h he, },
+end
```

Comment:
I suggest moving this to right after `is_cycle.not_of_nil`, which has a better `namespace` and `variables` set up for this particular lemma.

Here's a proof you can use that doesn't make use of `cons_is_cycle`
```lean
begin
  simp only [walk.is_cycle_def, walk.is_path_def, walk.is_trail_def, edges_cons, list.nodup_cons,
    support_cons, list.tail_cons],
  have : p.support.nodup → p.edges.nodup := edges_nodup_of_support_nodup,
  tauto,
end
```